test/Microsoft.AspNet.Mvc.Core.Test/Formatters/OutputFormatterTests.cs

```diff
@@ -102,6 +103,62 @@ public static IEnumerable<object[]> SelectResponseCharacterEncodingData
         }
 
         [Fact]
+        public void WriteResponse_GetMediaTypeWithCharsetReturnsMediaTypeFromCache_IfEncodingIsUtf8()
+        {
+            // Arrange
+            var formatter = new TestOutputFormatter();
+
+            var formatterContext = new OutputFormatterWriteContext(
+                new DefaultHttpContext(),
+                new TestHttpResponseStreamWriterFactory().CreateWriter,
+                objectType: null,
+                @object: null)
+            {
+                ContentType = new StringSegment("application/json"),
+            };
+
+            formatter.SupportedMediaTypes.Add("application/json");
+            formatter.SupportedEncodings.Add(Encoding.UTF8);
+
+            // Act
+            formatter.WriteAsync(formatterContext);
+            var firstContentType = formatterContext.ContentType;
+
+            formatterContext.ContentType = new StringSegment("application/json");
+
+            formatter.WriteAsync(formatterContext);
+            var secondContentType = formatterContext.ContentType;
+
+            // Assert
+            Assert.Same(firstContentType.Buffer, secondContentType.Buffer);
+        }
+
+        [Fact]
+        public void WriteResponse_GetMediaTypeWithCharsetReplacesCharset_IfDifferentThanEncoding()
+        {
+            // Arrange
+            var formatter = new TestOutputFormatter();
+
+            var formatterContext = new OutputFormatterWriteContext(
+                new DefaultHttpContext(),
+                new TestHttpResponseStreamWriterFactory().CreateWriter,
+                objectType: null,
+                @object: null)
+            {
+                ContentType = new StringSegment("application/json; charset=utf-7"),
+            };
+
+            formatter.SupportedMediaTypes.Add("application/json");
+            formatter.SupportedEncodings.Add(Encoding.UTF8);
```

Comment:
Actually, I need to do it this way to force the code path I'm looking for. Maybe that's an indication that we should rethink the scenario, but I think the scenario is valid, given that formatters are not selected using the charset, once the formatter has been selected, if it doesn't support the charset on the content type, it will pick the first charset it supports.